[SPARK-14409][ML] Adding a RankingEvaluator to ML

yongtang opened the pull request:
## What changes were proposed in this pull request?

This patch adds the implementation of Mean Rreciprocal Rank (MRR) in mllib.evaluation, and adds a RankingEvaluator to ML (ml.evaluation)
## How was this patch tested?

Additional test cast has been added to cover Mean Rreciprocal Rank (MRR).

yongtang committed:
[SPARK-14409][ML][WIP] Adding a RankingEvaluator to ML

This patch tries to add the implementation of Mean Rreciprocal Rank (MRR)
in mllib.evaluation, as a first step toward adding a RankingEvaluator to ML.

Additional test cast has been added to cover Mean Rreciprocal Rank (MRR).

This patch is a work in progress.

MLnick commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/RankingMetrics.scala:
```diff
@@ -156,6 +156,32 @@ class RankingMetrics[T: ClassTag](predictionAndLabels: RDD[(Array[T], Array[T])]
   }
 
+  /**
+   * Returns the mean reciprocal rank (MRR) of all the queries.
```
It would be good to expand the doc string here with more detail about what is MRR and how it is computed (see the `ndcg` doc string above).

yongtang replied:
Thanks @MLnick I just updated the docs.

AmplabJenkins commented:
Can one of the admins verify this patch?

MLnick commented:
ok to test

SparkQA commented:
**[Test build #56105 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56105/consoleFull)** for PR 12461 at commit [`2838c24`](https://github.com/apache/spark/commit/2838c240252e7a0578377d360f1a62e00bb48507).

SparkQA commented:
**[Test build #56105 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56105/consoleFull)** for PR 12461 at commit [`2838c24`](https://github.com/apache/spark/commit/2838c240252e7a0578377d360f1a62e00bb48507).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56105/
Test PASSed.

yongtang committed:
[SPARK-14409][ML][WIP] Adding a RankingEvaluator to ML

This patch tries to add the implementation of RankingEvaluator in ml.evaluation.

Additional test cast has been added to cover the implementation.

yongtang committed:
[SPARK-14409][ML][WIP] Adding a RankingEvaluator to ML

This patch tries to update the comment for mean reciprocal rank.

SparkQA commented:
**[Test build #56317 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56317/consoleFull)** for PR 12461 at commit [`af75c66`](https://github.com/apache/spark/commit/af75c66091770cc89d0846a761b6b296c3080f36).

yongtang commented:
@MLnick I added the implementation for ml.evaluation.RankingEvaluator. Please let me know if there are any issues.

SparkQA commented:
**[Test build #56317 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56317/consoleFull)** for PR 12461 at commit [`af75c66`](https://github.com/apache/spark/commit/af75c66091770cc89d0846a761b6b296c3080f36).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56317/
Test PASSed.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala:
```diff
@@ -0,0 +1,287 @@
+    val labelColName = $(labelCol)
+    val labelType = schema($(labelCol)).dataType
+    require(predictionType == labelType,
+      s"Prediction column $predictionColName and Label column $labelColName " +
```
these columns must in fact be `ArrayType`

MLnick commented on mllib/src/test/scala/org/apache/spark/ml/evaluation/RankingEvaluatorSuite.scala:
```diff
@@ -0,0 +1,60 @@
+        (Array[Int](1, 6, 2, 7, 8, 3, 9, 10, 4, 5), Array[Int](1, 2, 3, 4, 5)),
+        (Array[Int](4, 1, 5, 6, 2, 7, 3, 8, 9, 10), Array[Int](1, 2, 3)),
+        (Array[Int](1, 2, 3, 4, 5), Array[Int]())
+      ), 2)).toDF(Seq("prediction", "label"): _*).as[(Array[Int], Array[Int])]
```
If you pass in a DataFrame (i.e. without the `as[...]` call), do you get an exception? Predictors will still return `DataFrame`s and that DF may be passed into the evaluator. It may be safer to do the `as` call within the evaluator.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala:
```diff
@@ -0,0 +1,287 @@
+
+/**
+ * :: Experimental ::
+ * Evaluator for ranking, which expects two input columns: prediction and label.
```
I think we need a little more documentation here, perhaps mentioning the exact input types of these columns (as they're not just Double, they're Arrays, and this differs from the other evaluators).

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala:
```diff
@@ -0,0 +1,287 @@
+   * If a query has an empty ground truth set, the average precision will be zero and a log
+   * warning is generated.
+   */
+  private def meanAveragePrecision(dataset: Dataset[_]): Double = {
```
Originally the plan for the JIRA was to wrap the `RankingMetrics` in `mllib`. I'd prefer not to have to duplicate code. Can we wrap `RankingMetrics` here first? And then in a follow up we can look at moving the evaluators into `ml` and have the `mllib` versions wrap those (as has been done with many models etc).

yongtang replied:
Thanks @MLnick. I was trying to wrap the RankingMetrics. I encountered some issues when trying to cast an `Dataset[_]` into `RDD[(Array<T>, Array<T>)] `. Other Models typically have the exact type information (like `Double`) so it is not a problem.

Let me spend some more time on it and see if I can find a way to remove the code duplication.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala:
```diff
@@ -0,0 +1,287 @@
+   *
+   * IR evaluation methods for retrieving highly relevant documents. K. Jarvelin and J. Kekalainen
+   */
+  private def meanAveragePrecisionAtK(dataset: Dataset[_]): Double = {
```
I noticed that this is technically *not* the Mean Average Precision that is commonly referred to (e.g. [Kaggle](https://www.kaggle.com/wiki/MeanAveragePrecision)).

This metric is actually the average, of the precision at k, over the query set.

The MAP definition referenced is the average, of the *average* precision at k, over the query set. Our `meanAveragePrecision` method above is sort of a MAP@infinity. This `meanAveragePrecisionAtK` is actually something sort of like the Mean Precision at K.

I think we need the amend `meanAveragePrecision` to be MAP@k, and either ditch this method (since it is not really that much used vs MAP@k, or rename it)

cc @srowen @jkbradley

srowen replied:
Yes, it isn't quite MAP. It's the mean of precision @ k, rather than the mean of average precision @ k. And that is what you get by summing precision@i from 1 to k, but only for the i where item i is a hit/relevant. So when the labSet contains the item, add something like cnt/i to a running sum. Divide the sum at the end by k, instead of cnt.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala:
```diff
@@ -0,0 +1,287 @@
+            dcg += gain
+          }
+          if (i < labSetSize) {
+            maxDcg += gain
```
I also believe that the computation of NDCG might be incorrect. The "ideal" DCG (IDCG) as referenced [here](https://en.wikipedia.org/wiki/Information_retrieval#Discounted_cumulative_gain) and [here](https://www.kaggle.com/wiki/NormalizedDiscountedCumulativeGain) for example, seems to be computed as the best DCG that is possible given the *result* set (not the ground truth set). i.e. the max DCG possible when re-ordering the predicted set by true relevance (as opposed to predicted relevance).

srowen replied:
Relevance here is 0/1. It doesn't matter if you calculate the ideal DCG to the number of all relevant items, or all the way to n, since everything after the number of relevant items has relevance 0 and does not add to the sum.

MLnick replied:
I've come across implementations which compute IDCG based on a perfect re-ordering of the **prediction set**, rather than the **ground truth**. This seems based only on the ground truth set.

e.g, let's say we have 4 docs predicted with ground truth of > 4 docs, and the relevance is [1, 1, 0, 1]. Then DCG = `1.0 / log(2) + 1.0 / log(3) + 0.0 + 1.0 / log(5) = 2.974`. Under this formulation the IDCG = `1.0 / log(2) + 1.0 / log(3) + 1.0 / log(4) + 1.0 / log(5) = 3.696`. Thus NDCG = `0.805`.

However, if IDCG is the best possible DCG achievable **given the result set**, then IDCG = `1.0 / log(2) + 1.0 / log(3) + 1.0 / log(4) + 0.0 = 3.074` based on a best possible ordering of [1, 1, 1, 0]. Here NDCG = `0.967`.

srowen replied:
My understanding is the former. I mean, if the result set contained no relevant items, you'd have an nDCG of 0/0; if it manages to have just one relevant item but put it first, you'd return nDCG of 1. Those don't quite sound right. The wikipedia reference is ambiguous and the Kaggle link I think proposes the former, which at least in these cases would return 0 and something low, as perhaps expected.

MLnick replied:
Here is an example from [eBay](http://www.ebaytechblog.com/2010/11/10/measuring-search-relevance/) illustrating the latter.

> In our previous example, we had 2, then 0, 3, and a 2. The best arrangement of these same results would have been: 3, 2, 2, 0, that is, if the “great” result had been ranked first, followed by the two “relevant” ones, and then the “irrelevant”. This best ranking would have a DCG score of 3 / 1 + 2 / 2 + 2 / 3 + 0 / 4 = 4.67. This is known as the “ideal DCG,” or iDCG.

MLnick replied:
Also, the original paper referenced in the doc string seems to support the latter interpretation (bottom of pg 42 / top of pg 43). e.g.
> They do not unrealistically assume as a baseline that all retrieved documents could be maximally relevant. 
I don't think it's clear cut either way - but I definitely see good examples of the latter interpretation.

srowen replied:
Yeah the counter-argument is that, in the binary case, the best result set is almost surely all relevant, so this is only normalizing away the discount factor to put this on a scale of 0 to 1, and maybe make results comparable across different k. I thought that was its purpose, really.

The latter interpretation is strictly evaluating ranking within the returned results, and so doesn't measure whether the algorithm returned good results to begin with (doesn't measure anything about how the returned results are ranked relative to everything else). Failing to return a relevant result doesn't penalize you, which seems wrong.

The former interpretation is measuring something about both, which is why that version had always made sense to me as a metric

MLnick replied:
Yes fair enough. I don't think it's worth changing... and the doc actually does make it clear that it's using the DCG on the ground truth set.

MLnick commented on mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala:
```diff
@@ -0,0 +1,287 @@
+    val sqlContext = SQLContext.getOrCreate(sc)
+    import sqlContext.implicits._
+
+    dataset.map{ case (prediction: Array[T], label: Array[T]) =>
```
This may easily fail if the input dataset doesn't only have these columns. Typically above we would do a `select($(predictionCol), $(labelCol))` call.

yongtang commented:
Thanks @MLnick @srowen let me take some time and address all the issues.

yongtang committed:
[SPARK-14409][ML] Adding a RankingEvaluator to ML

This patch tries to consolidate ml.evaluation and mllib.evaluation
so that RankingEvaluator wraps RankingMetrics.

SparkQA commented:
**[Test build #56952 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56952/consoleFull)** for PR 12461 at commit [`a35d961`](https://github.com/apache/spark/commit/a35d9612c03cc48dae2ec2b06c0a0752d1a47919).

yongtang commented:
@MLnick @srowen I just updated the pull request to wrap the RankingEvaluator into calling RankingMetrics. Was finally able to fix the exception issue I previously encountered. Sorry for the delay and let me know if there are any issues.

SparkQA commented:
**[Test build #56952 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56952/consoleFull)** for PR 12461 at commit [`a35d961`](https://github.com/apache/spark/commit/a35d9612c03cc48dae2ec2b06c0a0752d1a47919).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56952/
Test PASSed.

SparkQA commented:
**[Test build #57506 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57506/consoleFull)** for PR 12461 at commit [`c50c469`](https://github.com/apache/spark/commit/c50c469ba2687f242785a1e6f4bdef6c4f69c45b).

SparkQA commented:
**[Test build #57506 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57506/consoleFull)** for PR 12461 at commit [`c50c469`](https://github.com/apache/spark/commit/c50c469ba2687f242785a1e6f4bdef6c4f69c45b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57506/
Test PASSed.

SparkQA commented:
**[Test build #73804 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73804/testReport)** for PR 12461 at commit [`78f13ad`](https://github.com/apache/spark/commit/78f13adf5a8b750ceea827e904f6b1761304f887).

SparkQA commented:
**[Test build #73804 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73804/testReport)** for PR 12461 at commit [`78f13ad`](https://github.com/apache/spark/commit/78f13adf5a8b750ceea827e904f6b1761304f887).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73804/
Test PASSed.

yongtang commented:
/cc @daniloascione please take a look.

AmplabJenkins commented:
Can one of the admins verify this patch?

flolas commented:
Hi, is there any plans for implementing RankingEvaluator?. I would be glad to help if needed.

Thanks.
